be/src/service/service_be/lake_service.cpp

```diff
@@ -429,4 +435,245 @@ void LakeServiceImpl::unlock_tablet_metadata(::google::protobuf::RpcController*
+    std::map<string, string> broker_prop(request->broker_properties().begin(), request->broker_properties().end());
+
+    // 1. get broker client
+    std::unique_ptr<BrokerServiceConnection> client;
+    std::unique_ptr<FileSystem> fs;
+    Status status = Status::OK();
+    client = std::make_unique<BrokerServiceConnection>(_env->broker_client_cache(), address, 10000, &status);
+    if (!status.ok()) {
+        std::stringstream ss;
+        ss << "failed to get broker client. "
+           << "broker addr: " << request->broker() << ". msg: " << status.get_error_msg();
+        cntl->SetFailed(status.get_error_msg());
+        LOG(WARNING) << ss.str();
+        return;
+    }
+    LOG(INFO) << "begin to upload snapshot files. num: " << request->snapshots().size()
+              << ", broker addr: " << request->broker();
+
+    for (auto& [tablet_id, snapshot] : request->snapshots()) {
+        // TODO: support report logic
+
+        // 2.1 get existing files from remote path
+        std::map<std::string, FileStat> remote_files;
+        status = _get_existing_files_from_remote(*client, snapshot.path(), broker_prop, &remote_files);
+        if (!status.ok()) {
+            LOG(ERROR) << "failed to get existing files from remote " << snapshot.path();
+            cntl->SetFailed("failed to get existing files from remote");
+            return;
+        }
+
+        for (auto& tmp : remote_files) {
+            VLOG(2) << "get remote file: " << tmp.first << ", checksum: " << tmp.second.md5;
+        }
+
+        std::map<std::string, std::string> file_locations;
+        auto tablet = _env->lake_tablet_manager()->get_tablet(tablet_id);
+        if (!tablet.ok()) {
+            LOG(ERROR) << "Fail to get tablet " << tablet_id;
+            cntl->SetFailed("Fail to get tablet");
+            return;
+        }
+        auto tablet_metadata = tablet->get_metadata(snapshot.version());
+        if (!tablet_metadata.ok()) {
+            LOG(ERROR) << "Fail to get tablet metadata" << tablet_id;
+            cntl->SetFailed("Fail to get tablet metadata");
+            return;
+        }
+        for (const auto& rowset : (*tablet_metadata)->rowsets()) {
+            for (std::string segment : rowset.segments()) {
+                file_locations[segment] = tablet->segment_location(segment);
+            }
+        }
+        file_locations[starrocks::lake::tablet_metadata_filename(tablet_id, snapshot.version())] =
+                tablet->metadata_location(snapshot.version());
+
+        for (auto& [file_name, file] : file_locations) {
+            auto md5sum = fs::md5sum(file);
+            if (!md5sum.ok()) {
+                LOG(ERROR) << "Fail to get md5 of file " << file;
+                cntl->SetFailed("Fail to get md5 of file");
+                return;
+            }
+
+            // check if this local file need upload
+            bool need_upload = false;
+            auto find = remote_files.find(file);
```

Comment:
not sure if this has the filepath together, lake table has a `staros://xxxx` style path location, not sure what it looks like from broker remote file listing. just doubt it here.

Reply:
> not sure if this has the filepath together, lake table has a `staros://xxxx` style path location, not sure what it looks like from broker remote file listing. just doubt it here.

I think we should use `file_name` here instead of `file`. I will fix it.